W-21311839-Add-gRPC-to-ACB-docs-CP (#600)
mulesoft/docs-code-builder: modules/ROOT/pages/des-design-grpc-api-spec.adoc

= Designing a gRPC API Specification with Anypoint Code Builder
:page-deployment-options: cloud-ide, desktop-ide

include::reuse::partial$beta-banner.adoc[tag="anypoint-code-builder"]

Anypoint Code Builder supports designing gRPC API specifications using Protocol Buffers (Protobuf 3). Create and edit `.proto` files, import reusable Protobuf libraries from Anypoint Exchange, and use syntax highlighting and validation.

> Use the Protocol Buffers 3 (Protobuf 3) specification language to design gRPC APIs. gRPC uses HTTP/2 and supports four RPC patterns:

Reviewer: ```suggestion
Anypoint Code Builder supports designing gRPC API specifications using Protocol Buffers 3 (Protobuf 3). Create and edit `.proto` files, import reusable Protobuf libraries from Anypoint Exchange, and use syntax highlighting and validation. gRPC uses HTTP/2 and supports four RPC patterns:
```
The first sentence repeats the intro.

Author: OK! Thanks!

This name is used as the API spec title in Exchange, the name of the spec file, and the name of the project's root directory.

include::partial$acb-reusable-steps.adoc[tags="no-project-name-reuse"]
| *Project Location* | Your home directory or another directory you create.

See xref:start-add-folders.adoc[].

include::partial$acb-reusable-steps.adoc[tags="proj-directory-warn"]
| *API Type* | The type of API spec to create.

Select *gRPC API* to create a gRPC project.

| *API Specification Language* | The language for the API specification.

Select *Protobuf 3* for gRPC API projects.
|===
. Click *Create Project*.
+
If prompted, trust the authors of the files in the folder.
+
When the project is ready, the API project opens with a default `.proto` file in the Editor view.
+
. Continue designing your gRPC API spec.
+
As you enter elements, use xref:start-discover-ui.adoc#auto-complete[auto-complete] (or press Ctrl+Space) to display available options within the context.

== Organize gRPC API Projects and Dependencies

A new gRPC API project includes a root `.proto` file (for example, `main.proto` or a file named after your project) that defines your service and messages.

> You edit the main `.proto` file to define `service`, `rpc`, and `message` blocks. To reuse shared types, add new `.proto` files in the same project and import them. You can also add a Protobuf fragment as a dependency from Exchange and import it in your specification.

Reviewer: ```suggestion
Edit the main `.proto` file to define `service`, `rpc`, and `message` blocks. To reuse shared types, add new `.proto` files in the same project and import them. You can also add a Protobuf fragment as a dependency from Exchange and import it in your specification.
```

Author: OK! Thanks!

== Publish a gRPC Fragment to Exchange

> You can publish reusable `.proto` definitions as a fragment to Anypoint Exchange so other gRPC API specifications can import them as dependencies.

Reviewer: ```suggestion
Publish reusable `.proto` definitions as a fragment to Anypoint Exchange so other gRPC API specifications can import them as dependencies.
```

Author: OK! Thanks!

A fragment contains shared definitions such as `message` or `enum` blocks.

=== Publish the Fragment

To publish a gRPC fragment to Exchange:

. Ensure that you're logged in to Anypoint Platform.
. Open the project that contains your reusable `.proto` files.
. From the IDE, select the publish action.
. When prompted:
+
* Select the business group.
* Provide the asset name and version.
* Add any required metadata.
. Confirm the publication.

After publishing, the fragment appears in Exchange as an API Fragment asset and can be added as a dependency to other gRPC API specification projects.

To use the fragment in another gRPC API specification:

. Add the fragment as a dependency from Exchange.
. Import the corresponding `.proto` file in your main specification.
. Validate that all references resolve correctly.

== Import a Protobuf Library from Anypoint Exchange

> You can import reusable Protobuf libraries (fragments) from Anypoint Exchange and add them as dependencies to your gRPC API project. You can reuse messages and services across multiple specifications.

Reviewer: ```suggestion
Import reusable Protobuf libraries (fragments) from Anypoint Exchange and add them as dependencies to your gRPC API project. You can reuse messages and services across multiple specifications.
```

Author: OK! Thanks!

To use a Protobuf fragment published to Exchange:

. Ensure you're logged in to Anypoint Platform.
. Open your gRPC API project.
. Open the Command Palette (Cmd+Shift+P on macOS or Ctrl+Shift+P on Windows).
. Select *MuleSoft: Add Fragment dependency from Exchange*.
. Search for the Protobuf fragment you need and select the version to add.
. After the dependency is added, the library is available under the project.
. In your main `.proto` file, use `import` to reference the library's package and message or service definitions. For example:
+
[source,protobuf]
----
syntax = "proto3";

import "path/to/library.proto";

service MyService {
  rpc GetOrder(OrderRequest) returns (OrderResponse);
}
----
. Save the spec. Anypoint Code Builder resolves the dependency and validates references. You can reuse messages and services from the imported library across your gRPC specs.

== Define Technical Details for gRPC Specifications

> 

Reviewer: Can we add a short sentence to introduce this header?

Author: OK! Thanks!

= Designing a gRPC API Specification with Anypoint Code Builder
:page-deployment-options: cloud-ide, desktop-ide

include::reuse::partial$beta-banner.adoc[tag="anypoint-code-builder"]

> Anypoint Code Builder supports designing gRPC API specifications using Protocol Buffers (Protobuf 3). Create and edit `.proto` files, import reusable Protobuf libraries from Anypoint Exchange, and use syntax highlighting and validation. When you design or publish your spec, the IDE generates a binary descriptor (`.protobin`) so Mule runtime can consume the service metadata while building the Mule application.

Reviewer 2: The binary descriptor (protobin) file is only generated when API is scaffolded or when gRPC APIspec is imported from exchange using the Command Import from Exchange.

Author: OK! Thanks!

Use the Protocol Buffers 3 (Protobuf 3) specification language to design gRPC APIs. gRPC uses HTTP/2 and supports four RPC patterns:

* Unary
* Client streaming
* Server streaming
* Bidirectional streaming

You author service and message definitions in `.proto` files and can reuse shared types by importing Protobuf fragments (libraries) from Exchange.

== Before You Begin

* xref:start-acb.adoc[].

> * Understand the basics of gRPC and Protocol Buffers (Protobuf 3). See https://grpc.io/docs/what-is-grpc/introduction/[Introduction to gRPC].

Reviewer 3: Also consider adding [Proto 3 language guide](https://protobuf.dev/programming-guides/proto3/)

Author: OK! Thanks!

* Have some familiarity with xref:access-management::business-groups.adoc[].
+

> When you publish your API spec to Exchange from Anypoint Code Builder, the IDE requests the name of the business group. See xref:access-management::business-groups.adoc[].

Reviewer 4: When you publish the business group is only asked in case the user did not define one upon creation. Business group is selected at creation time.

Author: OK! Thanks!

== Create a New gRPC API Specification Project

To create a gRPC API spec project in Anypoint Code Builder:

include::partial$acb-reusable-steps.adoc[tags="open-ide"]

. From *Create*, click *Design an API*:
. Complete the *API Specification* form:
+
[%header,cols="20a,60a"]
|===
| Field Name | Field Value

| *Project Name* | Unique name for your project.

This name is used as the API spec title in Exchange, the name of the spec file, and the name of the project's root directory.

include::partial$acb-reusable-steps.adoc[tags="no-project-name-reuse"]
| *Project Location* | Your home directory or another directory you create.

See xref:start-add-folders.adoc[].

include::partial$acb-reusable-steps.adoc[tags="proj-directory-warn"]
| *API Type* | The type of API spec to create.

Select *gRPC API* to create a gRPC project.

> | *Specification Language* | The language for the API specification.

Reviewer 3: Field name is _API Specification Language_

Author: OK! Thanks!

Select *Protobuf 3* for gRPC API projects.
|===
. Click *Create Project*.
+
If prompted, trust the authors of the files in the folder.
+
When the project is ready, the API project opens with a default `.proto` file in the Editor view.
+
. Continue designing your gRPC API spec.
+
As you enter elements, use xref:start-discover-ui.adoc#auto-complete[auto-complete] (or press Ctrl+Space) to display available options within the context.

== Organize gRPC API Projects and Dependencies

A new gRPC API project includes:

* A root `.proto` file (for example, `main.proto` or a file named after your project) that defines your service and messages.

> * A `protobuf-libraries` directory (or equivalent) for storing and organizing Protobuf fragments (reusable libraries) that you add as dependencies.

Reviewer 3: This is not included as part of the generated template.

Author: OK! Thanks!

> * A `protobuf-libraries` directory (or equivalent) for storing and organizing Protobuf fragments (reusable libraries) that you add as dependencies.

Reviewer 4: There is no protobuf-libraries directory for fragments. Fragments are added as dependencies and are directly accessible through the import statements. There is no directory created by default nor the user is required to create one

Author: OK! Thanks!

> You edit the main `.proto` file to define `service`, `rpc`, and `message` blocks. Dependencies from Exchange or local fragments are resolved from the `protobuf-libraries` area so you can reference shared message types and services across specs.

Reviewer 4: No protobuf-libraries is required. You can define any new .proto file within the same project and import it, or add a dependency to the project through the command palette and then import it.
No need for a protobuf-library area

Author: OK! Thanks!

== Publish a gRPC Fragment to Exchange

You can publish reusable `.proto` definitions as a fragment to Anypoint Exchange so other gRPC API specifications can import them as dependencies.

> A fragment contains shared definitions such as `message`, `enum`, or `service` blocks.

Reviewer 3: Just to clarify, there is no validation for fragment with service block so it is valid, but it is weird having services inside of a  fragment.

Reviewer 4: Agree, would remove service from here since we do not allow reusability of services

Author: OK! Thanks!

== Define Technical Details for gRPC Specifications

=== Declare Communication Patterns for gRPC Services

Anypoint Code Builder and Mule runtime support all four gRPC RPC patterns:

* **Unary** – Single request, single response.
* **Client streaming** – Client sends a stream of messages; server returns a single response.
* **Server streaming** – Client sends a single request; server returns a stream of messages.
* **Bidirectional streaming** – Both client and server send and receive streams of messages.

Define these in your `.proto` file using the appropriate `rpc` declarations and stream modifiers.

=== Verify Specification Accuracy with Syntax Validation

Anypoint Code Builder provides syntax highlighting and validation for Protobuf 3 specifications. Validation is performed using the API Modeling Framework (AMF) parser, so you get inline feedback and error reporting as you edit your `.proto` files. Use the editor and the xref:start-open-output-panel.adoc[output panel] to track validation and scaffolding status.

== Generate Binary Descriptors for Implementation

> During the design or publish process, Anypoint Code Builder generates a binary descriptor file (`.protobin`) from your Protobuf specification. With this descriptor, Mule runtime can understand the gRPC service metadata (services, methods, and message types) so you can implement and run the API. The generation typically occurs when you save, build, or publish the API spec. Don't create the `.protobin` file manually.

Reviewer 2: Binary Descriptor is only generated during scaffolding or Import Asset from Exchange command.

Reviewer 4: Exactly, this should be moved to the implementation documentation

Author: OK! Thanks!
